test/override-types.test-d.ts

```diff
@@ -0,0 +1,125 @@
+import { expectType } from 'tsd'
+import { TypeEqual } from 'ts-expect'
+import { PostgrestClient } from '../src'
+import { CustomUserDataType, Database } from './types'
+
+const REST_URL = 'http://localhost:54321'
+const postgrest = new PostgrestClient<Database>(REST_URL)
+
+// Test merge array result to object should error
+{
+  const singleResult = await postgrest
+    .from('users')
+    .select()
+    .overrideTypes<{ custom_field: string }>()
+  if (singleResult.error) {
+    throw new Error(singleResult.error.message)
+  }
+  let result: typeof singleResult.data
+  expectType<
+    TypeEqual<
+      typeof result,
+      {
+        Error: 'Type mismatch: Cannot cast array result to a single object. Use .returns<Array<YourType>> for array results or .single() to convert the result to a single object'
+      }
+    >
+  >(true)
+}
+
+// Test merge object result to array type should error
+{
+  const singleResult = await postgrest
+    .from('users')
+    .select()
+    .single()
+    .overrideTypes<{ custom_field: string }[]>()
+  if (singleResult.error) {
+    throw new Error(singleResult.error.message)
+  }
+  let result: typeof singleResult.data
+  expectType<
+    TypeEqual<
+      typeof result,
+      {
+        Error: 'Type mismatch: Cannot cast single object to array type. Remove Array wrapper from return type or make sure you are not using .single() up in the calling chain'
+      }
+    >
+  >(true)
+}
+
+// Test with single() / maybeSingle()
+{
+  const singleResult = await postgrest
+    .from('users')
+    .select()
+    .single()
+    .overrideTypes<{ custom_field: string }>()
+  if (singleResult.error) {
+    throw new Error(singleResult.error.message)
+  }
+  let result: typeof singleResult.data
+  expectType<TypeEqual<(typeof result)['custom_field'], string>>(true)
+}
+// Test with maybeSingle()
+{
+  const maybeSingleResult = await postgrest
+    .from('users')
+    .select()
+    .maybeSingle()
+    .overrideTypes<{ custom_field: string }>()
+  if (maybeSingleResult.error) {
+    throw new Error(maybeSingleResult.error.message)
+  }
+  let maybeSingleResultType: typeof maybeSingleResult.data
+  expectType<TypeEqual<(typeof maybeSingleResultType)['custom_field'], string>>(true)
```

Comment:
I think this should preserve optionality of the result when overriding.

Reply:
Yeah I tough about this as well.

I was worried about not being able to discriminate between an user opt-in optionality with `maybeSingle` and a value that can just be null and that the user might want to override.

But looking into this further, I guess the top level is always either an array of objects, a non-null object (`single`), or a nullable object (`maybeSingle`). If there is no other cases where the result can be null than the `maybeSingle` we can preserve optionality 🤔